pandas/_libs/algos.pyx

```diff
@@ -797,212 +797,202 @@ def rank_1d(
-                    are_diff(sorted_data[i + 1], val) or
-                    i == non_na_idx
-                )
+                next_val_diff = are_diff(masked_vals[_as[i]], masked_vals[_as[i+1]])
             else:
-                condition = (
-                    i == n - 1 or
-                    sorted_data[i + 1] != val or
-                    i == non_na_idx
-                )
-
-            if condition:
-
-                if tiebreak == TIEBREAK_AVERAGE:
-                    for j in range(i - dups + 1, i + 1):
-                        ranks[argsorted[j]] = sum_ranks / dups
-                elif tiebreak == TIEBREAK_MIN:
-                    for j in range(i - dups + 1, i + 1):
-                        ranks[argsorted[j]] = i - dups + 2
-                elif tiebreak == TIEBREAK_MAX:
-                    for j in range(i - dups + 1, i + 1):
-                        ranks[argsorted[j]] = i + 1
-                elif tiebreak == TIEBREAK_FIRST:
-                    if rank_t is object:
-                        raise ValueError('first not supported for non-numeric data')
+                next_val_diff = masked_vals[_as[i]] != masked_vals[_as[i+1]]
+        else:
+            next_val_diff = True
+
+        if (next_val_diff
+                or (mask[_as[i]] ^ mask[_as[i+1]])
+                or (check_labels and (labels_[_as[i]] != labels_[_as[i+1]]))
+        ):
+            # if keep_na, check for missing values and assign back
+            # to the result where appropriate
+            if keep_na and mask[_as[i]]:
+                for j in range(i - dups + 1, i + 1):
+                    out[_as[j]] = NaN
+                    grp_na_count = dups
+            elif tiebreak == TIEBREAK_AVERAGE:
+                for j in range(i - dups + 1, i + 1):
+                    out[_as[j]] = sum_ranks / <float64_t>dups
+            elif tiebreak == TIEBREAK_MIN:
+                for j in range(i - dups + 1, i + 1):
+                    out[_as[j]] = i - grp_start - dups + 2
+            elif tiebreak == TIEBREAK_MAX:
+                for j in range(i - dups + 1, i + 1):
+                    out[_as[j]] = i - grp_start + 1
+            elif tiebreak == TIEBREAK_FIRST:
+                for j in range(i - dups + 1, i + 1):
+                    if ascending:
+                        out[_as[j]] = j + 1 - grp_start
                     else:
-                        for j in range(i - dups + 1, i + 1):
-                            ranks[argsorted[j]] = j + 1
-                elif tiebreak == TIEBREAK_FIRST_DESCENDING:
-                    for j in range(i - dups + 1, i + 1):
-                        ranks[argsorted[j]] = 2 * i - j - dups + 2
-                elif tiebreak == TIEBREAK_DENSE:
-                    total_tie_count += 1
-                    for j in range(i - dups + 1, i + 1):
-                        ranks[argsorted[j]] = total_tie_count
-                sum_ranks = dups = 0
-
-    else:
-        with nogil:
```

Comment:
The original `rank_1d` looked like:
```
if object:
  rank()
else:
  with nogil:
     rank()
```
Something similar could be done here, but I first wanted to ask if using `with nogil:` actually helps here (since this algorithm is purely sequential).

Reply:
it's not about speed (though it can be) rather about releasing gil whenever possible; this way higher level multi processing can be used

Reply:
Added gil block